fdbclient/NativeAPI.actor.cpp

```diff
@@ -2243,6 +2243,45 @@ void Transaction::atomicOp(const KeyRef& key, const ValueRef& operand, MutationR
 	TEST(true); //NativeAPI atomic operation
 }
 
+ACTOR Future<Void> executeCoordinators(DatabaseContext* cx, StringRef execPayload, Optional<UID> debugID) {
+	try {
+		if (debugID.present()) {
+			g_traceBatch.addEvent("TransactionDebug", debugID.get().first(), "NativeAPI.executeCoordinators.Before");
+		}
+
+		state ExecRequest req(execPayload, debugID);
+		if (debugID.present()) {
+			g_traceBatch.addEvent("TransactionDebug", debugID.get().first(),
+									"NativeAPI.executeCoordinators.Inside loop");
+		}
+		wait(loadBalance(cx->getMasterProxies(), &MasterProxyInterface::execReq, req, cx->taskID));
+		if (debugID.present())
+			g_traceBatch.addEvent("TransactionDebug", debugID.get().first(),
+									"NativeAPI.executeCoordinators.After");
+		return Void();
+	} catch (Error& e) {
+		TraceEvent("NativeAPI.executeCoordinatorsError").error(e);
+		throw;
+	}
+}
+
+void Transaction::execute(const KeyRef& cmdType, const ValueRef& cmdPayload) {
+	TraceEvent("Execute operation").detail("Key", cmdType.toString()).detail("Value", cmdPayload.toString());
+
+	if (cmdType.size() > CLIENT_KNOBS->KEY_SIZE_LIMIT) throw key_too_large();
+	if (cmdPayload.size() > CLIENT_KNOBS->VALUE_SIZE_LIMIT) throw value_too_large();
+
+	auto& req = tr;
+
+	// Helps with quickly finding the exec op in a tlog batch
+	setOption(FDBTransactionOptions::FIRST_IN_BATCH);
+
+	auto& t = req.transaction;
+	auto r = singleKeyRange(cmdType, req.arena);
+	auto v = ValueRef(req.arena, cmdPayload);
+	t.mutations.push_back(req.arena, MutationRef(MutationRef::Exec, r.begin, v));
```

Comment:
The rest of your code won't correctly handle if there's >1 ExecOps per transaction, so you'll probably want to ban that here.

It's also worth putting some thought into if ExecOps can be submitted in the same transaction as other mutations.  e.g. if you think `tr.set('foo', 'bar') ; tr.execute('baz')` is a thing a client should be able to do

...I'm still actually unclear if you resolved the "what if multiple transactions use FIRST_IN_BATCH in the same batch" issue?  I'm still of the opinion that this would all be easier if you just sent TLog ExecOps separately from the mutation stream.

Reply:
> The rest of your code won't correctly handle if there's >1 ExecOps per transaction, so you'll probably want to ban that here.
Yes, aware of that.

> It's also worth putting some thought into if ExecOps can be submitted in the same transaction as other mutations. e.g. if you think tr.set('foo', 'bar') ; tr.execute('baz') is a thing a client should be able to do
Batch will contains transactions of exec and others - since that is allowed and tested because of running snapshots with read/write workload etc, i think this should also work just fine. Is there anything that concerns you in this area? I could not think of any.

> ...I'm still actually unclear if you resolved the "what if multiple transactions use FIRST_IN_BATCH in the same batch" issue? I'm still of the opinion that this would all be easier if you just sent TLog ExecOps separately from the mutation stream.

Yes, I tested with multiple FIRST_IN_BATCH and it happily puts two transactions in the same batch - so essentially the FIRST_IN_BATCH is not honored. Yes, the current logic does not work if there are multiple exec ops in the same batch. So, my options are as follows:
- send the exec ops as a separate payload in tlog commit request
- enforce FIRST_IN_BATCH
- make TLog code to work with multiple exec ops in the same batch

My only thing against payload in tlog commit request is, how will we handle the case when we want to snapshot only the remote region or both local and remote region. That case involves parsing the tlog mutations, so the current solution will be a generic solution.